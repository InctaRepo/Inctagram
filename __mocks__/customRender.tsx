/* eslint-disable */
import React, { ReactElement } from 'react'
import { Provider } from 'react-redux'

import { store } from '@/store'
import { render, renderHook, RenderOptions } from '@testing-library/react'

export const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    // <ThemeProvider theme="light">
    //   <TranslationProvider messages={defaultStrings}>
    <Provider store={store}>{children}</Provider>
    //   </TranslationProvider>
    // </ThemeProvider>
  )
}

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AllTheProviders, ...options })

export const customRenderHook = (hook: any, options?: Omit<RenderOptions, 'wrapper'>) => {
  return renderHook(() => hook(), { wrapper: AllTheProviders, ...options })
}
export * from '@testing-library/react'
export * from '@testing-library/user-event'
// eslint-disable-next-line import/export
export { customRender as render }
