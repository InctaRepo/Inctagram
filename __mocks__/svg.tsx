import React, { SVGProps, forwardRef } from 'react'

const SvgrMock = forwardRef<SVGSVGElement, SVGProps<SVGSVGElement>>((props, ref) => (
  <svg ref={ref} {...props} />
))

export const ReactComponent = SvgrMock
export default SvgrMock
