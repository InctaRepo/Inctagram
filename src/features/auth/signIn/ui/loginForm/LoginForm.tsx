import React from 'react'

import { useLoginForm } from '@/features/auth/signIn/hooks'
import { authByGitHub, authByGoogle } from '@/features/auth/successGoogleGitHub'
import GithubIcon from '@/public/icon/gitHubIcon.svg'
import GoogleIcon from '@/public/icon/googleIcon.svg'
import { RouteNames } from '@/shared/const'
import { LoginFormType } from '@/shared/schemas/createLoginSchema'
import { Button } from '@/ui/button'
import { Card } from '@/ui/card'
import { ControlledTextField } from '@/ui/controlled'
import { Typography } from '@/ui/typography'
import { DevTool } from '@hookform/devtools'
import Link from 'next/link'

import s from '@/features/auth/signIn/ui/loginForm/loginForm.module.scss'

type Props = {
  errorServer?: string
  onSubmitHandler?: (data: LoginFormType) => void
}
export const LoginForm = ({ errorServer, onSubmitHandler }: Props) => {
  const { control, errors, handleSubmit, router, submitData, t } = useLoginForm({
    errorServer,
    onSubmitHandler,
  })

  return (
    <Card className={s.card}>
      <div className={s.content}>
        <Typography className={s.title} variant={'h1'}>
          {t.auth.signIn}
        </Typography>
        <div className={s.oauthWrap}>
          <div onClick={authByGoogle}>
            <GoogleIcon />
          </div>
          <div onClick={authByGitHub}>
            <GithubIcon />
          </div>
        </div>
        <form className={s.form} onSubmit={handleSubmit(submitData)}>
          <DevTool control={control} />
          <ControlledTextField
            autoComplete={'username'}
            className={`${s.field} ${errors.email && s.fieldWithError && errorServer}`}
            control={control}
            fullWidth
            label={t.auth.email}
            name={'email'}
          />

          <ControlledTextField
            autoComplete={'current-password'}
            className={`${s.passField} ${errors.password && s.fieldWithError && errorServer}`}
            control={control}
            fullWidth
            label={t.auth.password}
            name={'password'}
            type={'password'}
          />
          <div className={s.wrapLinkForgotPass}>
            <Link className={s.link} href={RouteNames.PASSWORD_RECOVERY}>
              <Typography className={s.linkForgotPass} variant={'medium14'}>
                {t.auth.forgotPassword}
              </Typography>
            </Link>
          </div>
          <Button className={s.singIn} fullWidth type={'submit'} variant={'primary'}>
            <Typography variant={'bold16'}>{t.auth.signIn}</Typography>
          </Button>
        </form>

        <Typography variant={'regular16'}>{t.auth.dontHaveAccount}</Typography>
        <Button color={'link'} onClick={() => router.push(RouteNames.SIGN_UP)} variant={'link'}>
          {t.auth.signUp}
        </Button>
      </div>
    </Card>
  )
}
