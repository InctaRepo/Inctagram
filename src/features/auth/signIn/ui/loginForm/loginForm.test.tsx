import React from 'react'

import { customRender as render } from '@/__mocks__/customRender'
import { screen, waitFor } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'

import { LoginForm } from './LoginForm'

jest.mock('next/router', () => ({
  useRouter: jest.fn().mockReturnValue({ locale: 'en' }),
}))
const onSubmitHandler = jest.fn()

function setup(jsx: React.JSX.Element) {
  return {
    user: userEvent.setup(),
    ...render(jsx),
  }
}

describe('LoginForm', () => {
  it('renders LoginForm component', async () => {
    render(<LoginForm />)

    expect(screen.getByLabelText('Email')).toBeInTheDocument()
    expect(screen.getByRole('email', { name: /email/i })).toBeInTheDocument()
    expect(screen.getByLabelText('Password')).toBeInTheDocument()
    expect(screen.getByRole('password', { name: /password/i })).toBeInTheDocument()
    expect(screen.getByText('Forgot password?')).toBeInTheDocument()
    expect(screen.getByRole('link', { name: /Forgot password?/i })).toBeInTheDocument()
    expect(screen.getByText('Sign In', { selector: 'p.bold16.primary' })).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /sign in/i })).toBeInTheDocument()
    expect(screen.getByText('Sign Up')).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /sign up/i })).toBeInTheDocument()
  })

  it('calls onSubmitHandler with form data on submit', async () => {
    const { debug, user } = setup(<LoginForm onSubmitHandler={onSubmitHandler} />)

    await user.type(screen.getByRole('email', { name: /email/i }), 'test@example.com')
    await user.type(screen.getByRole('password', { name: /password/i }), '1qaz@WSX')
    await user.click(screen.getByRole('button', { name: /sign in/i }))
    expect(onSubmitHandler).toHaveBeenCalledTimes(1)
    expect(onSubmitHandler).toHaveBeenCalledWith({
      email: 'test@example.com',
      password: '1qaz@WSX',
    })
  })
  it('Min number of characters', async () => {
    const { debug, user } = setup(<LoginForm onSubmitHandler={onSubmitHandler} />)

    await user.type(screen.getByRole('email', { name: /email/i }), 'test@example.com')
    await user.type(screen.getByRole('password', { name: /password/i }), '123')
    await user.click(screen.getByRole('button', { name: /sign in/i }))
    expect(screen.getByText('Min number of characters 6')).toBeInTheDocument()
    expect(onSubmitHandler).not.toHaveBeenCalledTimes(1)
  })
  it('Invalid email address', async () => {
    const { debug, user } = setup(<LoginForm onSubmitHandler={onSubmitHandler} />)

    await user.type(screen.getByRole('email', { name: /email/i }), 'test')
    await user.type(screen.getByRole('password', { name: /password/i }), '123456')
    await user.click(screen.getByRole('button', { name: /sign in/i }))
    expect(screen.getByText('Invalid email address')).toBeInTheDocument()
    expect(onSubmitHandler).not.toHaveBeenCalledTimes(1)
  })

  it('displays error message when form submission fails', async () => {
    render(<LoginForm errorServer={'Invalid email or password'} />)

    await waitFor(() => {
      const errorElements = screen.queryAllByText(/Invalid email or password/i, {
        selector: 'p.regular14.primary.errorWrap',
      })

      expect(errorElements.length).toBeGreaterThan(0)
    })
  })
  it('snapshot LoginForm', () => {
    const snapshot = render(<LoginForm />)

    waitFor(() => expect(snapshot).toMatchSnapshot())
  })
})
