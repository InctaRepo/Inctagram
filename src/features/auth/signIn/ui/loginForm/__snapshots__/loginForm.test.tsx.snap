// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm snapshot LoginForm 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card card"
      >
        <div
          class="content"
        >
          <p
            class="h1 primary  title"
          >
            Sign In
          </p>
          <div
            class="oauthWrap"
          >
            <div>
              <svg />
            </div>
            <div>
              <svg />
            </div>
          </div>
          <form
            class="form"
          >
            <div
              class="field undefined fullWidth"
            >
              <div
                class="textFieldWrap"
              >
                <label
                  class="label undefined"
                  data-testid="email"
                  for="email"
                >
                  Email
                </label>
                <input
                  autocomplete="username"
                  class="textField undefined undefined"
                  id="email"
                  name="email"
                  role="email"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="passField undefined fullWidth"
            >
              <div
                class="textFieldWrap"
              >
                <label
                  class="label undefined"
                  data-testid="password"
                  for="password"
                >
                  Password
                </label>
                <input
                  autocomplete="current-password"
                  class="textField undefined undefined"
                  id="password"
                  name="password"
                  role="password"
                  type="password"
                  value=""
                />
                <svg
                  class="iconEye undefined"
                />
              </div>
            </div>
            <div
              class="wrapLinkForgotPass"
            >
              <a
                class="link"
                href="/auth/password-recovery"
              >
                <p
                  class="medium14 primary  linkForgotPass"
                >
                  Forgot password?
                </p>
              </a>
            </div>
            <button
              class="primary fullWidth singIn"
              type="submit"
            >
              <p
                class="bold16 primary  undefined"
              >
                Sign In
              </p>
            </button>
          </form>
          <p
            class="regular16 primary  undefined"
          >
            Don’t have an account?
          </p>
          <button
            class="link  undefined"
            color="link"
          >
            Sign Up
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card card"
    >
      <div
        class="content"
      >
        <p
          class="h1 primary  title"
        >
          Sign In
        </p>
        <div
          class="oauthWrap"
        >
          <div>
            <svg />
          </div>
          <div>
            <svg />
          </div>
        </div>
        <form
          class="form"
        >
          <div
            class="field undefined fullWidth"
          >
            <div
              class="textFieldWrap"
            >
              <label
                class="label undefined"
                data-testid="email"
                for="email"
              >
                Email
              </label>
              <input
                autocomplete="username"
                class="textField undefined undefined"
                id="email"
                name="email"
                role="email"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="passField undefined fullWidth"
          >
            <div
              class="textFieldWrap"
            >
              <label
                class="label undefined"
                data-testid="password"
                for="password"
              >
                Password
              </label>
              <input
                autocomplete="current-password"
                class="textField undefined undefined"
                id="password"
                name="password"
                role="password"
                type="password"
                value=""
              />
              <svg
                class="iconEye undefined"
              />
            </div>
          </div>
          <div
            class="wrapLinkForgotPass"
          >
            <a
              class="link"
              href="/auth/password-recovery"
            >
              <p
                class="medium14 primary  linkForgotPass"
              >
                Forgot password?
              </p>
            </a>
          </div>
          <button
            class="primary fullWidth singIn"
            type="submit"
          >
            <p
              class="bold16 primary  undefined"
            >
              Sign In
            </p>
          </button>
        </form>
        <p
          class="regular16 primary  undefined"
        >
          Don’t have an account?
        </p>
        <button
          class="link  undefined"
          color="link"
        >
          Sign Up
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
